libWlz/WlzCannyDeriche.3
'\" te
.\" ident MRC HGU $Id$
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" Project:    Woolz
.\" Title:      WlzCannyDeriche.3
.\" Date:       July 1999
.\" Author:     Bill Hill
.\" Copyright:	1999 Medical Research Council, UK.
.\"		All rights reserved.
.\" Address:	MRC Human Genetics Unit,
.\"		Western General Hospital,
.\"		Edinburgh, EH4 2XU, UK.
.\" Purpose:    Woolz function for Canny/Deriche edge detection.
.\" $Revision$
.\" Maintenance:Log changes below, with most recent at top of list.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.TH "WlzCannyDeriche" 3 "MRC HGU Woolz" "Woolz Procedure Library"
.SH NAME
WlzCannyDeriche \- Performs Canny/Deriche edge detection.
.SH SYNOPSIS
.LP
.B #include <Wlz.h>
.LP
.BI "WlzObject *WlzCannyDeriche(WlzObject **" dstGObj ,
.in +4m
.BI "WlzObject **" "srcObj" ,
.BI "double " "alpha" ,
.BI "double " "mult" ,
.br
.BI "WlzPixelV "pMinGrdV" ,
.BI "WlzPixelV "sMinGrdV" ,
.br
.BI "WlzErrorNum *" dstErr );
.in -4m
.SH ARGUMENTS
.TS
tab(^);
lI l.
dstGObj^destination pointer for gradient magnitude,
^may be NULL.
srcObj^given object
alpha^Deriche filter parameter.
mult^Filter multiplier.
pMinGrdV^primary hysteresis threshold value.
sMinGrdV^secondary hysteresis threshold value.
dstErr^destination error code pointer, may be NULL
.TE
.SH MT-LEVEL
.LP
Safe
.SH DESCRIPTION
Performs Canny/Deriche edge detection by the following steps:
.in +2m
1. A recursive first derivative Deriche filter is applied in
each dimension.
.br
2. The magnitude of gradient image is calculated.
.br
3. Non-maximal suppression is perform on the magnitude ofgradient image.
.br
4. A hysteresis threshold is applied to the Non-maximally suppressed image.
.in -2m
The returned object is the non-maximally suppressed edge
direction object
from WlzNMSuppress(3).
.SH EXAMPLES
.LP
.ps -2
.cs R 24
.nf

/*
 * Simple example of using WlzRsvFilter() to Gaussian smooth a 2D
 * or 3D domain object with grey values.
 */

#include <stdio.h>
#include <stdlib.h>
#include <Wlz.h>

int             main(char *argv[], int argc)
{
  WlzObject     *inObj = NULL,
                *outObj = NULL;
  WlzPixelV     min0,
                min1;
  WlzErrorNum   errNum = WLZ_ERR_NONE;

  min0.type = WLZ_GREY_INT;
  min0.v.inv = 8;
  min1.type = WLZ_GREY_INT;
  min1.v.inv = 32;
  if((inObj= WlzReadObj(stdin, &errNum)) == NULL)
  {
   (void )fprintf(stderr,
                  "%s: Failed to read input object.\\n",
                  *argv);
  }
  else if(((outObj = WlzCannyDeriche(NULL, inObj, 1.0,
                              1.0, min0, min1, &errNum)) == NULL) ||
      (errNum != WLZ_ERR_NONE))
  {
   (void )fprintf(stderr,
                  "%s: Failed to perrorm edge detection.\\n",
                  *argv);
  }
  else if((errNum = WlzWriteObj(stdout,
                            outObj)) != WLZ_ERR_NONE)
  {
   (void )fprintf(stderr,
                  "%s: Failed to write output object.\\n",
                  *argv);
  }
  if(outObj)
  {
   (void )WlzFreeObj(outObj);
  }
  if(inObj)
  {
   (void )WlzFreeObj(inObj);
  }
  return(errNum);
}

.fi
.cs R
.ps +2
.SH SEE ALSO
WlzGreyGradient(3), WlzSuppress(3), WlzRsvFilter(3),
WlzHyThreshold(3), WlzError(3)
.SH BUGS
Still to be found.


ChangeLog
$Log$
Revision 1.2  2007/12/05 13:33:30  bill
*** empty log message ***



binWlz/WlzCannyDeriche.c
#ifndef DOXYGEN_SHOULD_SKIP_THIS
#pragma ident "MRC HGU $Id$"
/***********************************************************************
* Project:      Woolz
* Title:        WlzRsvFilterObj.c
* Date:         March 1999
* Author:       Bill Hill
* Copyright:	1999 Medical Research Council, UK.
*		All rights reserved.
* Address:	MRC Human Genetics Unit,
*		Western General Hospital,
*		Edinburgh, EH4 2XU, UK.
* Purpose:      Canny/Deriche edge detection filter for Woolz domain
*		objects with grey values.
* $Revision$
* Maintenance:	Log changes below, with most recent at top of list.
************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <Wlz.h>

typedef enum
{
  WLZ_CANNYDERICHE_FLTACT_DIR,
  WLZ_CANNYDERICHE_FLTACT_EDG,
  WLZ_CANNYDERICHE_FLTACT_GRD
} WlzCannyDericheFilterAction;

extern int      getopt(int argc, char * const *argv, const char *optstring);
 
extern char     *optarg;
extern int      optind,
                opterr,
                optopt;

int             main(int argc, char **argv)
{
  int		option,
		ok = 1,
		usage = 0;
  WlzObject	*inObj = NULL,
  		*outObj = NULL,
		*grdObj = NULL,
		*edgObj = NULL;
  double	alpha = 1.0,
  		mult = 1.0;
  WlzPixelV	pThrV,
  		sThrV;
  WlzCannyDericheFilterAction action = WLZ_CANNYDERICHE_FLTACT_EDG;
  WlzErrorNum	errNum = WLZ_ERR_NONE;
  FILE		*fP = NULL;
  char 		*outFileStr,
  		*inObjFileStr;
  static char	optList[] = "ho:a:m:l:u:deg",
		outFileStrDef[] = "-",
  		inObjFileStrDef[] = "-";

  opterr = 0;
  outFileStr = outFileStrDef;
  inObjFileStr = inObjFileStrDef;
  pThrV.type = sThrV.type = WLZ_GREY_INT;
  pThrV.v.inv = sThrV.v.inv = 1.0;
  while(ok && ((option = getopt(argc, argv, optList)) != -1))
  {
    switch(option)
    {
      case 'o':
        outFileStr = optarg;
	break;
      case 'h':
        usage = 1;
	ok = 0;
	break;
      case 'a':
	if((optarg == NULL) || (sscanf(optarg, "%lg", &alpha) != 1))
	{
	  usage = 1;
	  ok = 0;
	}
        break;
      case 'd':
        action = WLZ_CANNYDERICHE_FLTACT_DIR;
	break;
      case 'e':
        action = WLZ_CANNYDERICHE_FLTACT_EDG;
	break;
      case 'g':
        action = WLZ_CANNYDERICHE_FLTACT_GRD;
        break;
      case 'm':
        if((optarg == NULL) || (sscanf(optarg, "%lg", &mult) != 1))
	{
	  usage = 1;
	  ok = 0;
	}
      case 'l':
        if((optarg == NULL) || (sscanf(optarg, "%d", &(sThrV.v.inv)) != 1))
	{
	  usage = 1;
	  ok = 0;
	}
	break;
      case 'u':
        if((optarg == NULL) || (sscanf(optarg, "%d", &(pThrV.v.inv)) != 1))
	{
	  usage = 1;
	  ok = 0;
	}
	break;
      default:
        usage = 1;
	ok = 0;
	break;
    }
  }
  if(ok)
  {
    if((inObjFileStr == NULL) || (*inObjFileStr == '\0') ||
       (outFileStr == NULL) || (*outFileStr == '\0'))
    {
      ok = 0;
      usage = 1;
    }
  }
  if(ok && (optind < argc))
  {
    if((optind + 1) != argc)
    {
      usage = 1;
      ok = 0;
    }
    else
    {
      inObjFileStr = *(argv + optind);
    }
  }
  if(ok)
  {
    if((inObjFileStr == NULL) ||
       (*inObjFileStr == '\0') ||
       ((fP = (strcmp(inObjFileStr, "-")?
	      fopen(inObjFileStr, "r"): stdin)) == NULL) ||
       ((inObj= WlzAssignObject(WlzReadObj(fP, &errNum), NULL)) == NULL) ||
       (errNum != WLZ_ERR_NONE))
    {
      ok = 0;
      (void )fprintf(stderr,
		     "%s: failed to read object from file %s\n",
		     *argv, inObjFileStr);
    }
    if(fP && strcmp(inObjFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(ok)
  {
    switch(action)
    {
      case WLZ_CANNYDERICHE_FLTACT_DIR:
	outObj = WlzCannyDeriche(NULL, inObj, alpha, mult,
				 pThrV, sThrV, &errNum);
        break;
      case WLZ_CANNYDERICHE_FLTACT_EDG:
	edgObj = WlzCannyDeriche(&grdObj, inObj, alpha, mult,
				 pThrV, sThrV, &errNum);
	if(errNum == WLZ_ERR_NONE)
	{
	  outObj = WlzMakeMain(edgObj->type, edgObj->domain, grdObj->values,
			       NULL, NULL, &errNum);
	}
	WlzFreeObj(edgObj);
	WlzFreeObj(grdObj);
        break;
      case WLZ_CANNYDERICHE_FLTACT_GRD:
	edgObj = WlzCannyDeriche(&outObj, inObj, alpha, mult,
				 pThrV, sThrV, &errNum);
	WlzFreeObj(edgObj);
        break;
    }
    if(errNum != WLZ_ERR_NONE)
    {
      ok = 0;
       (void )fprintf(stderr, "%s Failed to filter object (%s)\n",
       	              *argv, WlzStringFromErrorNum(errNum, NULL));
    }
  }
  if(ok)
  {
    if(((fP = (strcmp(outFileStr, "-")?
	      fopen(outFileStr, "w"):
	      stdout)) == NULL) ||
       (WlzWriteObj(fP, outObj) != WLZ_ERR_NONE))
    {
      ok = 0;
      (void )fprintf(stderr,
		     "%s: failed to write output object\n",
		     *argv);
    }
    if(fP && strcmp(outFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(inObj)
  {
    WlzFreeObj(inObj);
  }
  if(outObj)
  {
    WlzFreeObj(outObj);
  }
  if(usage)
  {
    (void )fprintf(stderr,
    "Usage: %s%sExample: %s%s",
    *argv,
    " [-o<output object>] [-h] [-o] [-a#] [-d] [-e] [-g]\n"
    "        [-m#] [-l#] [-u#] [<input object>]\n"
    "Options:\n"
    "  -h  Prints this usage information\n"
    "  -o  Output object file name.\n"
    "  -a  Deriche's alpha parameter.\n"
    "  -d  Output object with edge direction encoding.\n"
    "  -e  Output object with edge gradients.\n"
    "  -g  Output object with all gradients.\n"
    "  -m  Filter multiplier parameter.\n"
    "  -l  Lower hysteresis threshold value.\n"
    "  -u  Upper hysteresis threshold value.\n"
    "Applies a Canny/Deriche edge detection filter to a Woolz domain object\n"
    "with grey values.\n"
    "The input object is read from stdin and the filtered object is\n"
    "written to stdout unless the filenames are given.\n",
    *argv,
    " -e -a 1.0 -l 15 -u 40 -m 10 -o edges.wlz in.wlz\n"
    "The input Woolz object is read from in.wlz, and filtered using a\n"
    "Deriche edge operator with alpha=1.0, gradients are multiplied by\n"
    "10.0 before non-maximal suppression or hysteresis thresholding. The\n"
    "hysteresis threshold values of 15 and 40 are used to threshold the\n"
    "edge image which is written to edge.wlz.\n");
  }
  return(!ok);
}
#endif /* DOXYGEN_SHOULD_SKIP_THIS */
